# cds_insertion

cds\_insertion is intended to add CDS features to transcriptome annotations lacking them (referred to later as the "raw" transcriptome annotation) (e.g. a GTF file generated by a transcript assembler such as stringtie (https://ccb.jhu.edu/software/stringtie/).  cds\_insertion accomplishes this by taking as input the raw transcriptome annotation as well as a separate transcriptome annotation that includes CDS features (e.g. GENCODE https://www.gencodegenes.org/), identifying overlaps between the raw transcriptome exons and annotated CDS start codons, and then translating the raw transcriptome transcripts (_in silico_) from the overlapping annotated start codons. Future versions will accept a bed file of start codon coordinates in lieu of a full annotation. The idea is to restrict inserted CDS features to those derived from start codons to those for which there is strong evidence of utilization (e.g. conservation, known protein sequences, evidence from Ribo-seq or similar) rather than inserting another layer of arbitrary decisions by performing de novo CDS predictions. cds\_insertion additionally identifies a number of transcript properties derived from the presence/properties of the CDS such as the UTR lengths and the presence/absence of a premature termination codon.

## Basic Requirements

python 2.7  
Bedtools v>=2.27  

## Basic usage:

`python /path/to/cds\_insertion.py --transcript\_gtf /path/to/raw/transcriptome.gtf --transcript\_fasta /path/to/raw/transcriptome.fa --annotation\_gtf /path/to/cds\_annotated\_transcriptome.gtf --outdir /path/to/output/`

Note that this assumes that bedtools is in your PATH. The location of the bedtools executable can be made explicit by adding `--bedtools\_path /path/to/bedtools`

## Restrict to CCDS

The `--CCDS` flag will require that annotation CDS entries contain the text "CCDS".  This is intended for use with a GENCODE annotation GTF, and will have the affect of using only start codons from consensus CDS features (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5753299/).

`python /path/to/cds\_insertion.py --transcript\_gtf /path/to/raw/transcriptome.gtf --transcript\_fasta /path/to/raw/transcriptome.fa --annotation\_gtf /path/to/cds\_annotated\_transcriptome.gtf --outdir /path/to/output/ --CCDS`

## bigGenePred generation

By default cds\_insertion will output new GTF files with CDS entries, but for some visualization purposes it is nice to have bigGenePred files, which allow individual codon coloring when viewed on the UCSC genome browser. Generating the bigGenePred files requires the following UCSC utilities (by Jim Kent): `bigGenePred`, `gtfToGenePred`, `bedToBigBed`, and `genePredToBigGenePred`. These are available for linux and MacOS here: http://hgdownload.soe.ucsc.edu/admin/exe/. Provided those utilities are available in your `PATH`, the only necessary modifications to the above command are `--make\_bigBed`, `--bigGenePred\_as_path /path/to/bigGenePred.as` and `--chrNameLength_path /path/to/chrNameLength.txt`.  The `chrNameLength.txt` file is just a two-column tsv containing chromosome name in the first column and chromosome length in the second. Note that this is just the first two columns of a `.fai` file such as is generated by `samtools faidx`.  The `bigGenePred.as` file can be downloaded here: https://genome.ucsc.edu/goldenpath/help/bigGenePred.html.  

The bigGenePred generation can be achieved (alongside the normal output) as follows:

`python /path/to/cds\_insertion.py --transcript\_gtf /path/to/raw/transcriptome.gtf --transcript\_fasta /path/to/raw/transcriptome.fa --annotation\_gtf /path/to/cds\_annotated\_transcriptome.gtf --outdir /path/to/output/ --make\_bigBed --bigGenePred\_as_path /path/to/bigGenePred.as --chrNameLength_path /path/to/chrNameLength.txt`

Note that paths to the UCSC executables can be made explicit using `--gtfToGenePred\_path` and equivalent for the rest.



usage: cds_insertion.py [-h] --transcript_gtf TRANSCRIPT_GTF  
                        --transcript_fasta TRANSCRIPT_FASTA --annotation_gtf  
                        ANNOTATION_GTF [--CCDS] --outdir OUTDIR  
                        [--bigGenePred_as_path BIGGENEPRED_AS_PATH]  
                        [--gtfToGenePred_path GTFTOGENEPRED_PATH]  
                        [--genePredToBigGenePred_path GENEPREDTOBIGGENEPRED_PATH]  
                        [--bedToBigBed_path BEDTOBIGBED_PATH]  
                        [--bedtools_path BEDTOOLS_PATH]  
                        [--chrNameLength_path CHRNAMELENGTH_PATH]  
                        [--make_bigBed] [--ptc_dist PTC_DIST]  
  
optional arguments:  
  -h, --help            show this help message and exit  
  --transcript_gtf TRANSCRIPT_GTF  
                        Input transcript gtf filename  
  --transcript_fasta TRANSCRIPT_FASTA  
                        Input transcript fasta (assumes generation with  
                        gffread)  
  --annotation_gtf ANNOTATION_GTF  
                        Input gtf filename that contains CDS (preferably CCDS)  
                        features  
  --CCDS                If set, CDS used will be restricted to CCDS.  
  --outdir OUTDIR       Path for output files  
  --bigGenePred_as_path BIGGENEPRED_AS_PATH  
                        Path to bigGenePred.as file  
  --gtfToGenePred_path GTFTOGENEPRED_PATH  
                        Path to gtfToGenePred exe  
  --genePredToBigGenePred_path GENEPREDTOBIGGENEPRED_PATH  
                        Path to genePredToBigGenePred exe  
  --bedToBigBed_path BEDTOBIGBED_PATH  
                        Path to bedToBigBed exe  
  --bedtools_path BEDTOOLS_PATH  
                        Path to bedtools exe  
  --chrNameLength_path CHRNAMELENGTH_PATH  
                        Path to chrNameLength file  
  --make_bigBed         If set, will attempt to use Kent utils to output  
                        bigbed file from gtfs. Requires chrNameLength_path and  
                        bigGenePred_as_path to be set. If Kent utils are not  
                        in PATH, set the paths to the executables as well.  
  --ptc_dist PTC_DIST   Set minimum PTC distance required for transcript to be  
                        considered putative NMD substrate. default = 55  
